### PR TITLE
CFE-3431: Added unless attribute for custom promise types

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6463/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6463/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6463/

File: tests/acceptance/30_custom_promise_types/09_unless.cf [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    "$(G.testfile)-6"
+      delete => init_delete;
+    "$(G.testfile)-7"
+      delete => init_delete;
+    "$(G.testfile)-8"
+      delete => init_delete;
+    "$(G.testfile)-9"
+      delete => init_delete;
+    "$(G.testfile)-10"
+      delete => init_delete;
```
Comment (vpodzime) [inline tests/acceptance/30_custom_promise_types/09_unless.cf:39]: This could use an slist (with just the indices).

**Review by vpodzime — COMMENTED** [review]
I don't quite understand the naming convention in the acceptance tests, but I guess the new test should not start with *09*? Looks good to me otherwise. And thanks for the comprehensive test!

File: tests/acceptance/30_custom_promise_types/09_unless.cf [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+{
+    interpreter => "/bin/bash";
+    path => "$(this.promise_dirname)/example_module.sh";
+}
+
+bundle agent test
+{
+  meta:
+    "description" -> { "CFE-3391" }
+      string => "Test that custom promises work with if attribute";
```
Comment (larsewi) [inline tests/acceptance/30_custom_promise_types/09_unless.cf:60]:
```suggestion
      string => "Test that custom promises work with unless attribute";
```

File: tests/acceptance/30_custom_promise_types/09_unless.cf (hunk 1 above)
Comment (olehermanse) [reply tests/acceptance/30_custom_promise_types/09_unless.cf:39]: This is the same as the if test.

File: tests/acceptance/30_custom_promise_types/09_unless.cf (hunk 2 above)
Comment (olehermanse) [reply tests/acceptance/30_custom_promise_types/09_unless.cf]: Thanks, fixed.

**Comment by olehermanse** [comment]
> I don't quite understand the naming convention in the acceptance tests, but I guess the new test should not start with 09? Looks good to me otherwise.

@vpodzime Haha, race condition. I wrote the test before log level tests (9 and 10) were merged. Fixed now.

> And thanks for the comprehensive test!

I guess thanks to my past self, this was easy now, just a copy paste of the if test (inverted).

**Review by vpodzime — APPROVED** [review]
ACK

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/6464/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/6464/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-6464/

**Pull request merged**